maml/apps/pes/_lammps.py

```diff
@@ -445,45 +463,57 @@ def _setup(self):
 
         with open(input_file, "w") as f:
             f.write(input_template.format(write_restart=self.write_command, restart_file=self.restart_file))
-        with open("init.mod", "w") as f:
-            f.write(
-                init_template.format(
-                    deformation_size=self.deformation_size,
-                    jiggle=self.jiggle,
-                    maxiter=self.maxiter,
-                    maxeval=self.maxeval,
-                    lattice=self.lattice,
-                    alat=self.alat,
-                    num_species=self.num_species,
-                    atom_type=self.atom_type,
-                    masses="\n".join(["mass {} {}".format(i + 1, i + 1) for i in range(self.num_species)]),
-                )
-            )
+        if None in [self.alat, self.lattice, self.atom_type]:
+            read_data = ["box             tilt large", "read_data       data.static"]
+            with open("init.mod", "w") as f:
+                f.write("\n".join(init_template[:-12] + read_data))
+        else:
+            with open("init.mod", "w") as f:
+                f.writelines("\n".join(init_template))
         with open("potential.mod", "w") as f:
             f.write(potential_template.format(ff_settings="\n".join(ff_settings)))
         with open("displace.mod", "w") as f:
             f.write(displace_template.format(read_restart=self.read_command, restart_file=self.restart_file))
         return input_file
 
-    def calculate(self):
+    def calculate(self, structures=None):
         """
         Calculate the elastic constant given Potential class.
+        Args:
+            structures (list): A list of structures. If structures=None, the alat, lattice and atom_type
+                should not be None when initializing the calculator.
+
+        Returns:
+            List of elastic constants.
         """
-        with ScratchDir("."):
-            input_file = self._setup()
-            with subprocess.Popen([self.LMP_EXE, "-in", input_file], stdout=subprocess.PIPE) as p:
-                stdout = p.communicate()[0]
-                rc = p.returncode
-            if rc != 0:
-                error_msg = "LAMMPS exited with return code %d" % rc
-                msg = stdout.decode("utf-8").split("\n")[:-1]
-                try:
-                    error_line = [i for i, m in enumerate(msg) if m.startswith("ERROR")][0]
-                    error_msg += ", ".join(msg[error_line:])
-                except Exception:
-                    error_msg += msg[-1]
-                raise RuntimeError(error_msg)
-            result = self._parse()
+        if structures:
+            if None not in [self.alat, self.lattice, self.atom_type]:
+                raise ValueError(
+                    "To calculate elastic constants of input structures, at least one of "
+                    "the alat, lattice and atom_type should be None when initializing the calculator."
+                )
```

Comment:
This may not be needed. Whenever the user provides structures, the structures can just overwrite the default structure building mechanism.

Reply:
This is necessary because the ._setup() method needs to identify the selected way to take structures (by input structure list or input alat, lattice and atom_type arguments) and write the init.mod file accordingly. This is decided in line 466-472. 

Without this ValueError, the current code may setup input files for the default alat, lattice and atom_type arguments, even if structures are provided in the calculate method. 

To clarify this issue, I have modified the documentation of alat, lattice and atom_type arguments. But to achieve "Whenever the user provides structures, the structures can just overwrite the default structure building mechanism.", I think we have to add more arguments into the elastic calculator class, like "by_structure=False". I think this is no better than just setting one of alat, lattice and atom_type arguments to be zero, as both way users need to modify the default arguments of the calculator.

Reply:
I think the fundamental issue is that the default method uses the structure constructed within LAMMPS. I think we can just make the LAMMPS part consistent by constructing *all* structures in the python code. For example, when the user provides `alat`, `lattice` etc, we can just construct a pymatgen Structure in `calculate`, and use this structure in `super().calculate`. We need to change the `in.elastic` template though.

Reply:
I agree. I will modify this part as well as the part defining the returned results in the next commit. Thanks for the helpful suggestions!